File: shared/typescript-hono/src/db-routes.ts
// /db routes: per-database CRUD. The withRepository middleware resolves the
// :database path param to a shared UserRepository and stashes it on the context;
// every handler under /:database/* reads it back. Handlers call the shared zod
// schemas and repositories from @bench/shared — no DB logic lives here.

import {
  INTERNAL_ERROR,
  INVALID_JSON_BODY,
  makeError,
  NOT_FOUND,
  resolveRepository,
  type UserRepository,
  zCreateUser,
  zUpdateUser
} from "@bench/shared";
import { Hono, type MiddlewareHandler } from "hono";

type DbVariables = {
  repository: UserRepository;
};

const withRepository: MiddlewareHandler<{ Variables: DbVariables }> = async (c, next) => {
  const database = c.req.param("database");
  if (!database) return c.json(makeError(NOT_FOUND, "database parameter missing"), 404);

  const repository = resolveRepository(database);
  if (!repository) return c.json(makeError(NOT_FOUND, `unknown database type: ${database}`), 404);

  c.set("repository", repository);
  await next();
};

export function createDbRoutes(): Hono<{ Variables: DbVariables }> {
  const dbRoutes = new Hono<{ Variables: DbVariables }>();

  dbRoutes.get("/:database/health", async (c) => {
    const repository = resolveRepository(c.req.param("database"));
    if (!repository) return c.text("Service Unavailable", 503);

    const healthy = await repository.healthCheck().catch(() => false);
    return healthy ? c.text("OK") : c.text("Service Unavailable", 503);
  });

  dbRoutes.use("/:database/*", withRepository);

  dbRoutes.post("/:database/users", async (c) => {
    const repository = c.get("repository");

    let body: unknown;
    try {
      body = await c.req.json();
    } catch (err) {
      return c.json(makeError(INVALID_JSON_BODY, err), 400);
    }

    const parsed = zCreateUser.safeParse(body);
    if (!parsed.success) {
      return c.json(makeError(INVALID_JSON_BODY, parsed.error.message), 400);
    }

    try {
      const user = await repository.create(parsed.data);
      return c.json(user, 201);
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  dbRoutes.get("/:database/users/:id", async (c) => {
    const repository = c.get("repository");
    const id = c.req.param("id");

    try {
      const user = await repository.findById(id);
      if (!user) {
        return c.json(makeError(NOT_FOUND, `user with id ${id} not found`), 404);
      }
      return c.json(user);
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  dbRoutes.patch("/:database/users/:id", async (c) => {
    const repository = c.get("repository");
    const id = c.req.param("id");

    let body: unknown;
    try {
      body = await c.req.json();
    } catch (err) {
      return c.json(makeError(INVALID_JSON_BODY, err), 400);
    }

    const parsed = zUpdateUser.safeParse(body);
    if (!parsed.success) {
      return c.json(makeError(INVALID_JSON_BODY, parsed.error.message), 400);
    }

    try {
      const user = await repository.update(id, parsed.data);
      if (!user) {
        return c.json(makeError(NOT_FOUND, `user with id ${id} not found`), 404);
      }
      return c.json(user);
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  dbRoutes.delete("/:database/users/:id", async (c) => {
    const repository = c.get("repository");
    const id = c.req.param("id");

    try {
      const deleted = await repository.delete(id);
      if (!deleted) {
        return c.json(makeError(NOT_FOUND, `user with id ${id} not found`), 404);
      }
      return c.json({ success: true });
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  dbRoutes.delete("/:database/users", async (c) => {
    const repository = c.get("repository");
    try {
      await repository.deleteAll();
      return c.json({ success: true });
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  dbRoutes.delete("/:database/reset", async (c) => {
    const repository = c.get("repository");
    try {
      await repository.deleteAll();
      return c.json({ status: "ok" });
    } catch (err) {
      return c.json(makeError(INTERNAL_ERROR, err), 500);
    }
  });

  return dbRoutes;
}

File: servers/kt-ktor/src/main/kotlin/com/bench/ktor/routes/DbRoutes.kt
package com.bench.ktor.routes

import com.bench.ktor.ApiException
import com.bench.ktor.appJson
import com.bench.shared.Consts
import com.bench.shared.db.Repositories
import com.bench.shared.db.UserRepository
import com.bench.shared.model.CreateUser
import com.bench.shared.model.UpdateUser
import com.bench.shared.model.Validation
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.server.application.ApplicationCall
import io.ktor.server.request.receiveText
import io.ktor.server.response.respond
import io.ktor.server.response.respondText
import io.ktor.server.routing.Route
import io.ktor.server.routing.delete
import io.ktor.server.routing.get
import io.ktor.server.routing.patch
import io.ktor.server.routing.post
import io.ktor.server.routing.route
import io.ktor.server.util.getOrFail
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * The `/db/{database}` routes: per-backend health plus the user CRUD lifecycle. `{database}`
 * resolves to one of the four repositories; an unknown name is a
 * `404 {"error":"not found","details":"unknown database type: <db>"}` on CRUD and a
 * `503` text body on health (mirrors the other servers).
 */
fun Route.dbRoutes(repositories: Repositories) {
    route("/db/{database}") {
        get("/health") {
            val healthy =
                repositories
                    .resolve(call.parameters.getOrFail("database"))
                    ?.let { runCatching { it.healthCheck() }.getOrDefault(false) } ?: false
            if (healthy) {
                call.respondText("OK", ContentType.Text.Plain)
            } else {
                call.respondText("Service Unavailable", ContentType.Text.Plain, HttpStatusCode.ServiceUnavailable)
            }
        }
        post("/users") {
            val user = repositories.repo(call).create(decodeCreate(call))
            call.respond(HttpStatusCode.Created, user)
        }
        delete("/users") {
            repositories.repo(call).deleteAll()
            call.respond(buildJsonObject { put("success", true) })
        }
        get("/users/{id}") {
            val id = call.parameters.getOrFail("id")
            call.respond(repositories.repo(call).findById(id) ?: throw notFoundUser(id))
        }
        patch("/users/{id}") {
            val id = call.parameters.getOrFail("id")
            call.respond(repositories.repo(call).update(id, decodeUpdate(call)) ?: throw notFoundUser(id))
        }
        delete("/users/{id}") {
            val id = call.parameters.getOrFail("id")
            if (repositories.repo(call).delete(id)) {
                call.respond(buildJsonObject { put("success", true) })
            } else {
                throw notFoundUser(id)
            }
        }
        delete("/reset") {
            repositories.repo(call).deleteAll()
            call.respond(buildJsonObject { put("status", "ok") })
        }
    }
}

private fun Repositories.repo(call: ApplicationCall): UserRepository {
    val database = call.parameters.getOrFail("database")
    return resolve(database) ?: throw ApiException.NotFound("unknown database type: $database")
}

private fun notFoundUser(id: String) = ApiException.NotFound("user with id $id not found")

private suspend fun decodeCreate(call: ApplicationCall): CreateUser {
    val data =
        try {
            appJson.decodeFromString<CreateUser>(call.receiveText())
        } catch (e: SerializationException) {
            throw ApiException.BadRequest(Consts.ERR_INVALID_JSON, e.message, e)
        }
    Validation.validateCreate(data)?.let { throw ApiException.BadRequest(Consts.ERR_INVALID_JSON, it) }
    return data
}

private suspend fun decodeUpdate(call: ApplicationCall): UpdateUser {
    val data =
        try {
            appJson.decodeFromString<UpdateUser>(call.receiveText())
        } catch (e: SerializationException) {
            throw ApiException.BadRequest(Consts.ERR_INVALID_JSON, e.message, e)
        }
    Validation.validateUpdate(data)?.let { throw ApiException.BadRequest(Consts.ERR_INVALID_JSON, it) }
    return data
}
